### PR TITLE
Ad-Hoc Brute-Force for tiered hnsw index - [MOD-4631]

### PR DESCRIPTION
**Pull request by GuyAv46**
**Describe the changes in the pull request**

This PR includes the implementation of `getDistanceFrom` for tiered HNSW.
It includes optimization for single-value index (the much more common), which we should consider implementing separately for single and multi indexes.

**Mark if applicable**

- [ ] This PR introduces API changes
- [ ] This PR introduces serialization changes

### CODE REVIEW

File: src/VecSim/vec_sim_tiered_index.h [hunk 1]
```diff
@@ -113,3 +114,45 @@ VecSimTieredIndex<DataType, DistType>::topKQuery(const void *queryBlob, size_t k
...
+        return flat_dist;
+    }
+
+    // Try to get the distance from the Main index.
+    this->mainIndexGuard.lock_shared();
+    auto main_dist = this->index->getDistanceFrom(id, blob);
+    this->mainIndexGuard.unlock();
+
+    // Return the minimum distance that is not NaN.
+    return std::fmin(flat_dist, main_dist);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/VecSim/vec_sim_tiered_index.h:157]: Yep

File: tests/unit/test_hnsw_tiered.cpp [hunk 2]
```diff
@@ -898,3 +896,256 @@ TYPED_TEST(HNSWTieredIndexTest, MergeMulti) {
...
+        };
+        TieredHNSWParams tiered_hnsw_params = {.hnswParams = params, .tieredParams = tiered_params};
+        auto *tiered_index = reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(
+            HNSWFactory::NewTieredIndex(&tiered_hnsw_params, allocator));
+        // Set the created tiered index in the index external context.
+        index_ctx.index_strong_ref.reset(tiered_index);
+        EXPECT_EQ(index_ctx.index_strong_ref.use_count(), 1) << IS_MULTI;
+
+        // Launch the BG threads loop that takes jobs from the queue and executes them.
+        size_t THREAD_POOL_SIZE = 1;
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_hnsw_tiered.cpp:1089]: leftovers

File: tests/unit/test_hnsw_tiered.cpp [hunk 3]
```diff
@@ -898,3 +896,256 @@ TYPED_TEST(HNSWTieredIndexTest, MergeMulti) {
...
+            ASSERT_EQ(0, VecSimIndex_GetDistanceFrom(search_job->index, label, query));
+
+            search_job->successful_searches++;
+            delete search_job;
+        };
+
+        // Insert vectors in parallel to search.
+        for (size_t i = 0; i < n; i++) {
+            GenerateAndAddVector<TEST_DATA_T>(tiered_index, dim, i % n_labels, i);
+            auto query = (TEST_DATA_T *)allocator->allocate(dim * sizeof(TEST_DATA_T));
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_hnsw_tiered.cpp:1111]: yes. `SearchJobMock` takes ownership of the query.
I can make it take a copy instead

File: tests/unit/test_hnsw_tiered.cpp [hunk 4]
```diff
@@ -898,3 +896,256 @@ TYPED_TEST(HNSWTieredIndexTest, MergeMulti) {
...
+        }
+        std::atomic_int successful_searches(0);
+
+        auto parallel_adhoc_search = [](AsyncJob *job) {
+            auto *search_job = reinterpret_cast<SearchJobMock *>(job);
+            auto query = search_job->query;
+            size_t element = *(TEST_DATA_T *)query;
+            size_t label = element % search_job->n;
+
+            ASSERT_EQ(0, VecSimIndex_GetDistanceFrom(search_job->index, label, query));
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply tests/unit/test_hnsw_tiered.cpp:1102]: IMHO I don't think it's necessary, we didn't intend to check the distance function, and its the best way to validate we found the best vector (especially in multi)

File: tests/unit/test_hnsw_tiered.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_hnsw_tiered.cpp:1138]: Not necessary IMO...

File: tests/unit/test_hnsw_tiered.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (alonre24) [reply tests/unit/test_hnsw_tiered.cpp:1129]: OK, I respect your humble opinion.

**Review by alonre24 — APPROVED** [review]

File: src/VecSim/vec_sim_tiered_index.h [hunk 5]
```diff
@@ -31,8 +31,8 @@ class VecSimTieredIndex : public VecSimIndexInterface {
     void *memoryCtx;
     UpdateMemoryCB UpdateIndexMemory;
 
-    std::shared_mutex flatIndexGuard;
-    std::shared_mutex mainIndexGuard;
+    mutable std::shared_mutex flatIndexGuard;
+    mutable std::shared_mutex mainIndexGuard;
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_tiered_index.h:35]: why the change?

File: tests/unit/test_hnsw_tiered.cpp [hunk 6]
```diff
@@ -890,3 +906,264 @@ TYPED_TEST(HNSWTieredIndexTest, MergeMulti) {
...
+    VecSimIndex_AddVector(hnsw_index, vec3_1, 3);
+    VecSimIndex_AddVector(flat_index, vec3_2, 3);
+    VecSimIndex_AddVector(hnsw_index, vec3_3, 3);
+
+    VecSimIndex_AddVector(hnsw_index, vec4_1, 4);
+    VecSimIndex_AddVector(hnsw_index, vec4_2, 4);
+    VecSimIndex_AddVector(flat_index, vec4_2, 4);
+    VecSimIndex_AddVector(flat_index, vec4_3, 4);
+
+    // vec4 is not inserted to any index, simulating a non-existing vector.
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_tiered.cpp:1058]:
```suggestion
    // vec5 is not inserted to any index, simulating a non-existing vector.
```

**Review by DvirDukhan — COMMENTED** [review]
nice, very small comment

**Review by DvirDukhan — APPROVED** [review]

**Review by DvirDukhan — APPROVED** [review]

**Review by alonre24 — APPROVED** [review]
I prefer keeping it simple and splitting both safe and unsafe "get_distance_from" instead of using the sophisticated c++ stuff, but not critical...
